db/version_set.cc

```diff
@@ -1124,67 +1244,172 @@ void LevelIterator::Seek(const Slice& target) {
     file_iter_.SeekToLast();
+    if (range_tombstone_iter_) {
+      TrySetDeleteRangeSentinel(file_smallest_key(file_index_));
+    }
+  }
+  if (!to_return_sentinel_) {
+    SkipEmptyFileBackward();
   }
-  SkipEmptyFileBackward();
   CheckMayBeOutOfLowerBound();
 }
 
 void LevelIterator::Next() {
+  // For range tombstones, Next() call should update range_tombstone_iter_
+  // if file_iter_ changes. Maintain active_range_tombstones_ set
+  // correspondingly. Before advancing to next SST file, check boundary for
+  // possible tombstone sentinel key.
   assert(Valid());
-  file_iter_.Next();
-  SkipEmptyFileForward();
+  if (to_return_sentinel_) {
+    // file_iter_ was at end of file already if we were to return sentinel.
+    ClearSentinel();
+    SkipEmptyFileForward();
+  } else {
+    file_iter_.Next();
+    if (range_tombstone_iter_) {
+      TrySetDeleteRangeSentinel(file_largest_key(file_index_));
+    }
+    if (!to_return_sentinel_) {
+      SkipEmptyFileForward();
+    }
+  }
 }
 
 bool LevelIterator::NextAndGetResult(IterateResult* result) {
   assert(Valid());
-  bool is_valid = file_iter_.NextAndGetResult(result);
+  // file_iter_ is at EOF already when to_return_sentinel_
+  bool is_valid = !to_return_sentinel_ && file_iter_.NextAndGetResult(result);
+  bool prefix_sentinel = false;
   if (!is_valid) {
+    if (to_return_sentinel_) {
+      prefix_sentinel = prefix_sentinel_;
+      ClearSentinel();
+    } else if (range_tombstone_iter_) {
+      TrySetDeleteRangeSentinel(file_largest_key(file_index_));
+      if (to_return_sentinel_) {
+        result->key = sentinel_;
+        result->bound_check_result = IterBoundCheck::kUnknown;
+        result->value_prepared = true;
+        return true /* valid */;
+      }
+    }
     is_next_read_sequential_ = true;
-    SkipEmptyFileForward();
+    // SkipEmptyFileForward() should be called regardless of whether there is
+    // prefix
+    if (SkipEmptyFileForward() && prefix_sentinel) {
+      // Previous SST file was potentially seeked to the end of file
+      // (SkipEmptyFileForward() returns true). We check the first key in the
+      // new file to see if it is at another prefix. By prefix_sentinel we have
+      // prefix_extractor != nullptr and !read_options_.total_order_seek, by
+      // file_iter_.Valid() we have file_iter_.iter() != nullptr.
+      if (!read_options_.auto_prefix_mode && file_iter_.Valid() &&
+          !seek_prefix_.empty()) {
+        size_t ts_sz = user_comparator_.timestamp_size();
+        Slice file_user_key_without_ts =
+            ExtractUserKeyAndStripTimestamp(file_iter_.key(), ts_sz);
+        if (!prefix_extractor_->InDomain(file_user_key_without_ts) ||
+            user_comparator_.CompareWithoutTimestamp(
+                seek_prefix_, false,
+                prefix_extractor_->Transform(file_user_key_without_ts),
+                false) != 0) {
+          SetFileIterator(nullptr);
+        }
+      }
```

Comment:
I think both the code here and the ones in `Seek()` open the next file in `SkipEmptyFileForward()`. This code here is to mimic the ones in Seek() and the main benefit I see is the "stricter contract" mentioned in the pre-existing comments in `Seek()` (the comments near the end of `Seek()`). 
> will an upper level take care of checking whether the key went outside the target prefix

I think we only do so if `readoptions. prefix_same_as_start` is set, and the check is in DBIter. Otherwise, users are expected to see keys outside of target prefix once all keys within the prefix are exhausted.

Reply:
Yes, `LevelIterator::Next()` could enter the next file and return a key outside of prefix. This won't happen in the merging iterator path however, since all child iterators are wrapped by `IteratorWrapper` which only invoked `NextAndGetResult` (https://github.com/facebook/rocksdb/blob/b0048b673c61f6ec52bb80fccb69aa08e569693b/table/iterator_wrapper.h#L89).

Reply:
But is `LevelIterator::NextAndGetResult()` different from `LevelIterator::Next()` in that regard (possibly returning a key outside the prefix)?

Reply:
Also it does look like the stricter contract is satisfied even without this, because it only requires "If keys for the prefix is exhausted, it is either positioned to the next key after the prefix, or make the iterator invalid.". If the prefix seek landed on a range tombstone sentinel and the next file were outside the prefix, wouldn't the next point key found by NextAndGetResult() be the first key after the prefix?

Reply:
>  is LevelIterator::NextAndGetResult() different from LevelIterator::Next() in that regard 

You are right, `LevelIterator::Next*` could return a key outside the originally seeked prefix.

> If the prefix seek landed on a range tombstone sentinel and the next file were outside the prefix, wouldn't the next point key found by NextAndGetResult() be the first key after the prefix?

I'm not sure about this, and that was why I added this part of the code.
Say prefix is "aa", an SST file contains "ab", ["aa1", "aa2"), then Seek("aa") will land on the sentinel, but the next point key found by `NextAndGetResult()` is the first key in the next SSTfile, which does not satisfy the contract.

Reply:
OK I agree NextAndGetResult() could skip an arbitrary sequence of keys past the prefix without something like this. 

What if during `Seek()` we remember the index of the first file whose smallest key is outside our seek prefix, and then use that to limit how far `SkipEmptyFileForward()` can advance?

Reply:
Your way looks more efficient. I still wonder about the need for #5861 as it's unclear why the higher layers care what key past the prefix they are shown. Anyways let me finish reviewing assuming we need to continue following this contract.

Reply:
> Your way looks more efficient.

Or is it trivial? If we only need to do the search when prefix bloom gives a negative result, and we look at the next file first, it could be 0 or 1 additional key comparisons in most cases.

Reply:
If you mean the performance impact of this block of code, yes it should be relatively small: `prefix_sentinel` is set in very few cases, and is cleared in the `ClearSentinel()` call above.

Reply:
Sorry I can't seem to get past this part every time I try to review. The conceptual difficulty of the prefix sentinel logic feels too high in order to satisfy an API contract that doesn't have a clear need.

Is it accurate that `Seek()` is only at risk of jumping keys after the prefix on the first file it considers, when it calls `file_iter_.Seek()`? It looks like `Seek()` calls `file_iter_.SeekToFirst()` on later files via `SkipEmptyFileForward()`. The first valid `SeekToFirst()` should be the first point key after the first file. Then any jumped key should be in the first file (?).

If that holds, can `Seek()` set a flag to tell `SkipEmptyFileForward()` to not enter the next file when `file_iter_.Seek()` returns invalid and the immediate next file's smallest key has a different prefix from the seek prefix?

Reply:
I think it's accurate, and putting the logic into `SkipEmptyFileForward()` looks viable to me. I'll explore the implementation of this.

Reply:
Updated.